webserver/templates/user/scraper_api.js

```diff
@@ -0,0 +1,309 @@
+    return select(
+            isSpotifyURI,
+            map(
+                function(elem) { return elem.getAttribute("href") },
+                this.rootScrobbleElement.getElementsByTagName("a")
+               )
+            )[0];
+};
+
+Scrobble.prototype.asJSONSerializable = function () {
+    return {
+        "track_metadata": {
+            "track_name": this.trackName(),
+            "artist_name": this.artistName(),
+        },
+        "listened_at": this.scrobbledAt()
+
+    }
+};
+
+function extractlastFMIDFromURL(loveButtonURL) {
+    var parts = loveButtonURL.split("/");
+    return parts.slice(0, parts.length-2).join("/");
+}
+
+function encodeScrobbles(jsonstr) {
+    var scrobbles = JSON.parse(jsonstr);
+    scrobbles = scrobbles['recenttracks']['track'];
+
+    var parsedScrobbles = map(function(rawScrobble) {
+        var scrobble = new Scrobble(rawScrobble);
+        return scrobble.asJSONSerializable();
+    }, scrobbles);
+
+
+    var structure = {
+        "listen_type" : "import",
+        "payload"     : parsedScrobbles
+    };
+
+    return structure;
+}
+
+function getLastFMPage(page) {
+    function retry(reason) {
+        console.warn(reason + ' fetching last.fm page=' + page + ', retrying in 3s');
+        setTimeout(function () {
+            getLastFMPage(page);
+        }, 3000);
+    }
+
+    var url = "http://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&user={{ lastfm_username }}&api_key={{ lastfm_api_key }}&format=json&page=" + page;
+
+    var xhr = new XMLHttpRequest();
+    xhr.timeout = 10 * 1000; // 10 seconds
+    xhr.open("GET", encodeURI(url));
+    xhr.onload = function () {
+        if (/^2/.test(this.status)) {
+            reportPageAndGetNext(this.response);
+        } else if (/^5/.test(this.status)) {
+            retry('got ' + this.status);
+
+        } else {
+            // ignore 40x
+            pageDone();
+        }
+    };
+    xhr.ontimeout = function () {
+        retry('timeout');
+    };
+    xhr.onabort = function () {
+        // this should never happen
+        pageDone();
+    };
+    xhr.onerror = function () {
+        retry('error');
+    };
+    xhr.send();
+    xhr = null;
+}
+
+var version = "1.4";
+var page = 1;
+var numberOfPages = 1;
+if (document.querySelector('.pages') != null) {
+   numberOfPages = parseInt(document.getElementsByClassName("pages")[0].innerHTML.trim().split(" ")[3]);
+}
+
+var toReport = [];
+var numCompleted = 0;
+var activeSubmissions = 0;
+
+var timesReportScrobbles = 0;
+var timesGetPage = 0;
+var times4Error = 0;
+var times5Error = 0;
+
+function getPageCount() {
+    function getCountAgain(reason) {
+        console.warn(reason + ' fetching last.fm pagecount, retrying in 3s');
+        setTimeout(function () {
+            getPageCount();
+        }, 3000);
+    }
+    var url = "http://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&user={{ lastfm_username }}&api_key={{ lastfm_api_key }}&format=json&page=" + page;
```

Comment:
yeah. I thought of the same. But using a separate function for separate tasks would be a good idea. And extracting the page numbers from the first call would require use of some flag variables which i tried to avoid.